keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,354 @@
+
+The VPC Load Balancer supports TCP and HTTP, and it is possible to create TCP and HTTP listeners for the same LB instance. UDP is not supported. 
+
+The VPC LB pricing is time and data amount based, i.e. the number of protocols on the same LB instance does not affect it.
+
+The NLB supports TCP and UDP on the same NLB instance. The usage of NLB does not have pricing effects, it is part of the IKS basic package.
+
+Summary: once this feature is implemented IBM Cloud VPC LB can use TCP and HTTP ports from a single Service definition. NLB can use TCP and UDP ports from a single Service definition.
+
+#### OpenStack
+
+The OpenStack CPI supports TCP, UDP, HTTP(S) in Service definitions and can configure the Octavia listeners with the protocols defined in the Service.
+OpenStack Octavia supports TCP, UDP and HTTP(S) on listeners, an own listener must be configured for each protocol, and different listeners can be used on the same LB instance.
+
+There was a bug in Octavia versions <5.0.0: it was not possible to use the same port number (5e.g. 53) with different protocols (e.g. TCP and UDP) on the same LB instance. It has been fixed in 5.0.0, which is available since the "T" release of OpenStack.
+
+Summary: the OpenStack based clouds that use Octavia v2 as their LBaaS seems to support this feature once implemented. It is true that the "T" release is the newest one, so upgrade may take a while. On the other hand OpenStack documentation mentions, that a newer Octavia version can be used with previous releases of other OpenStack projects, i.e. it can be the case that the upgrade effort is on Octavia side in an OpenStack cloud. 
+Pricing is up to the pricing model of the OpenStack providers.
+
+#### Oracle Cloud
+
+Oracle Cloud supports TCP, HTTP(S) protocols in its LB solution. The Oracle CPI also supports the protocols in he K8s Service definitions.
+
+The pricing is based on time and capacity: https://www.oracle.com/cloud/networking/load-balancing-pricing.html I.e. the amount of protocols on a sinlge LB instance does not affect the pricing.
+
+#### Tencent Cloud
+
+The Tencent Cloud CPI supports TCP, UDP and HTTP(S) in Service definitions and can configure the CLB listeners with the protocols defined in the Service.
+The Tencent Cloud CLB supports TCP, UDP and HTTP(S) listeners, an own listener must be configured for each protocol. 
+The number of listeners does not affect CLB pricing. CLB pricing is time (day) based and not tied to the number of listeners.
+https://intl.cloud.tencent.com/document/product/214/8848
+
+A user can ask for an internal Load Balancer via a K8s Service definition that  has the annotation `service.kubernetes.io/qcloud-loadbalancer-internal-subnetid: subnet-xxxxxxxx`. Internal CLBs are free.
+
+
+### Risks and Mitigations
```

Comment:
I think we also need to talk about upgrade/downgrade scenarios here. If I upgrade to a version that supports mixed protocol and then I downgrade to one that doesn't, will any of the existing cloud providers handle this unexpectedly? Sounds like worse case is dangling LBs because the old provider could assume single protocol listeners?

Reply:
@andrewsykim I added some thoughts about upgrade/downgrade. Please check!